### PR TITLE
[GRPO] Log generation entropy

### LINKED ISSUE #3571: [GRPO] Entropy metric
### Feature request

Entropy metrics for the generation during training, averaged across the sequence.

### Motivation

This has been up for discussion previously, particularly after the release of the [DAPO paper](https://arxiv.org/pdf/2503.14476) but was never implemented. Policy entropy is a real-time barometer of exploration capacity. Logging entropy gives practitioners an early-warning gauge of whether downstream performance is plateauing. 

Recent RLVR work makes a strong case for tracking entropy.

1. [The Entropy Mechanism of Reinforcement Learning for Reasoning Language Models](https://arxiv.org/pdf/2505.22617)
2. [ProRL: Prolonged Reinforcement Learning Expands Reasoning Boundaries in Large Language Models](https://arxiv.org/pdf/2505.24864)
3. [Beyond the 80/20 Rule: High-Entropy Minority Tokens Drive Effective Reinforcement Learning for LLM Reasoning](https://arxiv.org/pdf/2506.01939)
4. [The Surprising Effectiveness of Negative Reinforcement](https://arxiv.org/pdf/2506.01347)

### Your contribution

Will PR

### PR DESCRIPTION
**Pull request by LeonEricsson**
# What does this PR do?

Log the average per-token entropy of the generations/completions (normalized over the local batch, BNPO-style). Since entropy is not computed by default, you need to set `log_entropy` in the config. However, given the importance of this metric, I'm open to the idea of always computing entropy–provided that it doesn't introduce noticeable overhead.

Fixes #3571

This PR also includes a minor refactoring of the _compute_loss function.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LeonEricsson** [comment]
Test run comparing the compute cost of calculating entropy. Currently, I'm only using a batch size of 12 and a maximum completion length of 400.

W B Chart 07_07_2025, 13_25_13

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1430,35 +1431,67 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
         logits_to_keep = completion_ids.size(1)  # we only need to compute the logits for the completion tokens
 
-        # Compute the entropy at each position in the completion
-        if self.token_entropy_percentile_threshold > 0.0:
-            logps_and_entropies = self._get_per_token_logps_and_entropies(
-                model, input_ids, attention_mask, logits_to_keep, compute_entropy=True
+        entropies = None
+        entropy_mask = None
+
+        compute_entropy = self.log_entropy or self.token_entropy_percentile_threshold > 0.0
```
Comment (pramodith) [inline trl/trainer/grpo_trainer.py]: Shouldn't we be using `torch.no_grad()`, if we're computing entropy just for logging purposes?

Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: yes!

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1430,35 +1431,64 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+        if compute_entropy_mask:
             entropy_mask = self._compute_entropy_mask(entropies, completion_mask)
-        else:
-            per_token_logps = self._get_per_token_logps_and_entropies(
-                model, input_ids, attention_mask, logits_to_keep
-            )["logps"]
-            entropy_mask = None
+
+        if self.log_entropy and entropies is not None:
+            with torch.no_grad():
```
Comment (pramodith) [inline trl/trainer/grpo_trainer.py]: I think it'd be more efficient if the `torch.no_grad()` context is here https://github.com/huggingface/trl/blob/640a9f39164ce3f9bbac7d325c1149ba023314e6/trl/trainer/grpo_trainer.py#L905-L907. It should theoretically be faster and have fewer memory ops.

Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: Yeah, I was considering this but didn’t want to complicate things with `_get_per_token_logps_and_entropies`. I’ll take another look and see if I can find a clean way to incorporate it. Either way, I’m leaning toward always logging entropy instead of making it configurable, which will probably change things.

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -927,7 +927,8 @@ def _get_per_token_logps_and_entropies(
             all_logps.append(logps)
 
             if compute_entropy:
-                entropies = entropy_from_logits(logits)
+                with torch.no_grad():
+                    entropies = entropy_from_logits(logits)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py:1099]: the entropies, as they are used by _compute_loss (masking and/or logging), don't require gradients. However, this may not be true for future use cases. Is this too naive?

@pramodith

**Comment by LeonEricsson** [comment]
Updated PR with new changes on main

File: trl/trainer/grpo_trainer.py (hunk 3 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py:1099]: I agree that the entropies don't need gradients, but when we compute the entropies for the entropy masking loss, the logits are involved in ops and I think these ops should be considered in the computational graph for gradient calculation. So I do think we need to differentiate between when we compute entropy for logging vs masking.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1537,6 +1538,9 @@ def _compute_loss(self, model, inputs):
             mean_kl = (per_token_kl * completion_mask).sum() / completion_mask.sum()
             self._metrics[mode]["kl"].append(self.accelerator.gather(mean_kl).nanmean().item())
 
+        mean_entropy = (entropies * completion_mask).sum() / completion_mask.sum()
```
Comment (pramodith) [inline trl/trainer/grpo_trainer.py]: I think we can run into an edge case here where completion_mask.sum() is 0, when someone uses `mask_truncated_completions`all of the logging steps should probably include a small offset in the denominator, like (completion_mask.sum + 1e-8) to prevent a _nan_ from popping up.

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -903,7 +903,7 @@ def _get_last_hidden_state(self, unwrapped_model, input_ids, attention_mask, log
     def _get_per_token_logps_and_entropies(
         self, model, input_ids, attention_mask, logits_to_keep, batch_size=None, compute_entropy=False
     ) -> dict[str, Optional[torch.Tensor]]:
-        """Compute log‐probs and (optionally) entropies for each token."""
+        """Compute log-probs and (optionally) entropies for each token."""
```
Comment (pramodith) [inline trl/trainer/grpo_trainer.py:1060]: Wondering if there's any benefit to having the `compute_entropy` bool, seems like it'd always be True and we don't seem to be providing an option to the user to set it.

**Review by pramodith — COMMENTED** [review]
Thanks for working on this! Had a few questions.

**Comment by qgallouedec** [comment]
Thanks! Reviewing asap

File: trl/trainer/grpo_trainer.py (hunk 5 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py:1060]: these calls for example have `compute_entropy=False`

https://github.com/huggingface/trl/blob/56f4201db6973b82cfe55efab9ac7e1d64b2677a/trl/trainer/grpo_trainer.py#L1560-L1597

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: take a look now and let me know what you think

File: trl/trainer/grpo_trainer.py (hunk 5 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py:1060]: My bad, I realised this a few hours later.

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: Looks great!

File: trl/trainer/grpo_trainer.py (hunk 3 above)
Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py:1099]: Since we only use entropy to mask tokens (a hard 0/1 gates) there’s no gradient path through that thresholding anyway, masking is non differentiable. So wrapping in torch.no_grad() should be safe for both masking and logging.

**Review by pramodith — APPROVED** [review]
LGTM for what its worth 😅
